src/helpers.rs

```diff
@@ -853,21 +853,46 @@ impl<'mir, 'tcx> Evaluator<'mir, 'tcx> {
 #[derive(Clone)]
 pub struct CurrentSpan<'a, 'mir, 'tcx> {
     span: Option<Span>,
+    tcx: TyCtxt<'tcx>,
     machine: &'a Evaluator<'mir, 'tcx>,
 }
 
 impl<'a, 'mir, 'tcx> CurrentSpan<'a, 'mir, 'tcx> {
+    /// Get the current span, skipping non-local frames.
+    /// This function is backed by a cache, and can be assumed to be very fast.
     pub fn get(&mut self) -> Span {
-        *self.span.get_or_insert_with(|| Self::current_span(self.machine))
+        *self.span.get_or_insert_with(|| Self::current_span(self.tcx, self.machine))
     }
 
+    /// Similar to `CurrentSpan::get`, but retrieves the parent frame of the first non-local frame.
+    /// This is useful when we are processing something which occurs on function-entry and we want
+    /// to point at the call to the function, not the function definition generally.
     #[inline(never)]
-    fn current_span(machine: &Evaluator<'_, '_>) -> Span {
+    pub fn get_parent(&mut self) -> Span {
+        let stack = self.machine.threads.active_thread_stack();
+        for i in (1..stack.len()).rev() {
+            // Skip #[track_caller] frames
+            if stack[i].instance.def.requires_caller_location(self.tcx) {
+                continue;
+            }
+            let def_id = stack[i].instance.def_id();
+            if def_id.is_local() || self.machine.local_crates.contains(&def_id.krate) {
```

Comment:
There's a lot of code duplication here with `current_span`. Is there any nice way to share that?